### PR TITLE
Wasm32-wasi target support

### PR DESCRIPTION
**Pull request by rjzak**
I'm working on getting Tokio working on Rust, and some changes were needed for Mio.

Companion to https://github.com/tokio-rs/tokio/pull/4716

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

File: src/poll.rs [hunk 1]
```diff
@@ -225,7 +226,7 @@ use std::{fmt, io};
 /// [`SourceFd`]: unix/struct.SourceFd.html
 /// [`Poll::poll`]: struct.Poll.html#method.poll
 pub struct Poll {
-    registry: Registry,
+    registry: Arc<Registry>,
```
Comment (carllerche) [inline src/poll.rs:229]: You will probably want to move this `Arc` into the wasi implementation of `Selector`.

File: src/poll.rs [hunk 2]
```diff
@@ -270,13 +271,21 @@ impl Poll {
...
-                registry: Registry { selector },
+                registry: Arc::new(Registry { selector }),
             })
         }
     }
 
     /// Create a separate `Registry` which can be used to register
     /// `event::Source`s.
+    #[cfg(target_os = "wasi")]
+    pub fn registry(&self) -> Arc<Registry> {
```
Comment (Thomasdezeeuw) [inline src/poll.rs:282]: We're not going to accept a different public API based on the compile target.

**Review by Thomasdezeeuw — CHANGES_REQUESTED** [review]
@rjzak we're not going to accept this as is. `Poll` needs to have mutable access on Windows as it can only handle a single poll call at a time. Furthermore we don't want to force everyone to use an `Arc`.

Could you please explain why you need this change? "Wasm improvements" are not really clear.

**Comment by rjzak** [comment]
@Thomasdezeeuw I certainly wouldn't want this PR accepted as-is, as the test application here https://github.com/rjzak/tokio-echo-test compiles to Wasm but doesn't work. I thought that by sharing the code I have so far, I could receive some feedback and pointers.

**Comment by Thomasdezeeuw** [comment]
Ping @rjzak.

**Comment by rjzak** [comment]
@Thomasdezeeuw I'm still working on it!

File: src/poll.rs (hunk 1 above)
Comment (rjzak) [reply src/poll.rs:229]: What would that look like? I just need a `Registry` object, not a reference, and not an object that gets "moved".

**Comment by rjzak** [comment]
@Thomasdezeeuw I have Axum & Tokio working on WebAssembly.
Still baffled at the moment with that to do with `Arc<Registry>`. This is where it's used in Tokio: https://github.com/rjzak/tokio/blob/wasi_wip/tokio/src/io/driver/mod.rs#L145. I could use some assistance.

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw I have Axum & Tokio working on WebAssembly. Still baffled at the moment with that to do with `Arc<Registry>`. This is where it's used in Tokio: https://github.com/rjzak/tokio/blob/wasi_wip/tokio/src/io/driver/mod.rs#L145. I could use some assistance.

Maybe you can put the [`Selector`](https://github.com/tokio-rs/mio/blob/c6f39d6ddba12b32793453dfe65a74104445f68d/src/sys/wasi/mod.rs#L43-L48) in an `Arc`?

**Comment by rjzak** [comment]
@Thomasdezeeuw @carllerche I moved the `Arc<>` to `Registry::selector`, and all seems to be okay, except for the tests on Windows. On Windows, the `Poll::poll()` function is failing with an error stating that a mutable reference is needed for `Arc<Selector>`, however, the function `Selector::select()` doesn't take a mutable reference to itself, so this *shouldn't* be a problem, unless I'm misunderstanding the error.

Line in question: https://github.com/tokio-rs/mio/blob/6312041389bc37343d1e81151a29eb8ac68e87ea/src/poll.rs#L378
Error:

    error[E0596]: cannot borrow data in an `Arc` as mutable
       --> src\poll.rs:378:9
        |
    378 |         self.registry.selector.select(events.sys(), timeout)
        |         ^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
        |
        = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::sync::Arc<sys::windows::selector::Selector>`

Otherwise, formatting issues and the previously API difference between platforms has been resolved.

File: src/sys/wasi/mod.rs [hunk 3]
```diff
@@ -110,6 +110,13 @@ impl Selector {
         }
     }
 
+    pub fn try_clone(&self) -> io::Result<Selector> {
```
Comment (Thomasdezeeuw) [inline src/sys/wasi/mod.rs:113]: Nit: `pub(crate)`.

File: src/poll.rs [hunk 4]
```diff
@@ -275,8 +275,7 @@ impl Poll {
         }
     }
 
-    /// Create a separate `Registry` which can be used to register
-    /// `event::Source`s.
+    /// Get access to the `Registry`, which can be used to register `event::Source`s.
```
Comment (Thomasdezeeuw) [inline src/poll.rs:278]: Nit: please revert.

**Review by Thomasdezeeuw — COMMENTED** [review]
Two small nits, LGTM otherwise.

File: src/poll.rs (hunk 4 above)
Comment (rjzak) [reply src/poll.rs]: Reverted

File: src/sys/wasi/mod.rs (hunk 3 above)
Comment (rjzak) [reply src/sys/wasi/mod.rs]: Changed

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
Thanks @rjzak.

**Comment by rjzak** [comment]
Thank you!

**Pull request merged**